### PR TITLE
Optimize `ecMul` precompile (scalar multiplication on BN254)

### PR DESCRIPTION
**Pull request by wborgeaud**
Use essentially the same optimizations as in #840. Instruction count goes from 38969 to 25837.
Relies on #840.

### CODE REVIEW

File: evm/src/cpu/kernel/asm/curve/bn254/curve_mul.asm [hunk 1]
```diff
@@ -19,77 +14,29 @@ global ec_mul:
...
+    PUSH bn_mul_after_a SWAP1 PUSH @SEGMENT_KERNEL_BN_WNAF_A PUSH @BN_SCALAR %jump(wnaf)
+global bn_mul_after_a:
+    // stack: b, x, y, bn_msm, bn_mul_after_glv_precompute_and_msm, retdest
+    PUSH bn_mul_after_b SWAP1 PUSH @SEGMENT_KERNEL_BN_WNAF_B PUSH @BN_SCALAR %jump(wnaf)
+global bn_mul_after_b:
+    // stack: x, y, bn_msm, bn_mul_end, retdest
+    %jump(bn_precompute_table)
+global bn_mul_end:
+    %stack (Ax, Ay, retdest) -> (retdest, Ax, Ay)
+    JUMP
```
Comment (unzvfu) [inline evm/src/cpu/kernel/asm/curve/bn254/curve_mul.asm:42]: nit: Newline missing at end of file.

File: evm/src/cpu/kernel/asm/curve/bn254/curve_mul.asm [hunk 2]
```diff
@@ -19,77 +14,29 @@ global ec_mul:
...
-    JUMP
-
-odd_scalar:
-    // stack: x', y', x, y, retdest
-    %jump(ec_add_valid_points)
+bn_mul_valid_point:
+    %stack (x, y, s, retdest) -> (s, bn_mul_after_glv, x, y, bn_msm, bn_mul_end, retdest)
+    %jump(bn_glv_decompose)
+global bn_mul_after_glv:
+    // stack: bneg, a, b, x, y, bn_msm, bn_mul_after_glv_precompute_and_msm, retdest
```
Comment (unzvfu) [inline evm/src/cpu/kernel/asm/curve/bn254/curve_mul.asm:29]: nit: `bn_mul_after_glv_precompute_and_msm` -> `bn_mul_end`
Idem on next two stack comments.

File: evm/src/cpu/kernel/asm/curve/bn254/curve_mul.asm [hunk 3]
```diff
@@ -19,77 +14,29 @@ global ec_mul:
...
+    %bn_check
     // stack: isValid(x, y), x, y, s, retdest
-    %jumpi(ec_mul_valid_point)
+    %jumpi(bn_mul_valid_point)
     // stack: x, y, s, retdest
     %pop3
-    %ec_invalid_input
+    %bn_invalid_input
 
 // Same algorithm as in `exp.asm`
```
Comment (unzvfu) [inline evm/src/cpu/kernel/asm/curve/bn254/curve_mul.asm:24]: nit: This should be updated or removed.

**Review by unzvfu — APPROVED** [review]
Looks good to me, just a few minor nits.

Something we missed in your previous PR: The global symbols in `curve/secp256k1/precomputation.asm` should have the `secp` prefix for consistency with those in `curve/bn254/precomputation.asm`.

File: evm/src/cpu/kernel/asm/curve/bn254/glv.asm [hunk 4]
```diff
@@ -0,0 +1,97 @@
+// Inspired by https://github.com/AztecProtocol/weierstrudel/blob/master/huff_modules/endomorphism.huff
```
[the remark this answers is not in the document]
Comment (wborgeaud) [reply evm/src/cpu/kernel/asm/curve/bn254/glv.asm:1]: Good point, I'll look into it and make the change in another PR.

**Pull request merged**